### DISCUSSION TITLE
Handling non-pen touch events

### DISCUSSION THREAD

**Discussion by casept** [flxzt/rnote#513] [Development]
I'm currently working on implementing a ruler tool. The way this tool works in most applications is that it can be moved with one finger and rotated with 2, while remaining stationary when interacted with by the pen to prevent accidental movement when trying to use it to draw a line.

How can I handle finger touches in particular? The `PenBehaviour` trait doesn't seem helpful, and a brief search of the codebase hasn't turned up anything useful, either.

**Comment by flxzt** [comment]
Currently this is not possible, because touch events are either filtered out completely before they reach the individual pens, or they are reported as `PenEvent`s when `touch-drawing` is enabled.

Handling touch is tricky, because of the navigation gestures installed on the `CanvasWrapper`, in particular the pinch gesture for zooming. Without having tested it I would guess it blocks propagation to the canvas completely as soon as it detects two touch points, and I don't think we can move it into the `bubble` phase where events are handled *after* the canvas had its chance. I think I tried that before and had some some other issues (although I don't remember the details).

So that would leave us with single-point touch events (for now). I think we have the option to do either:

- introduce a field in the pen events telling where they originated (stylus, mouse, touch) to be able to differentiate between them. we'd then need to be able to always handle touch events, and reject them on a per-pen level. So we'd need a return value back to the canvas widget telling it if the event should be propagated or not. The touch-drawing option would then be a preference of the brush.
- A simpler alternative: draw some handles / nodes so that the drags for moving / rotating the ruler have to originate from them. We'd probably need something like this anyway to enable usage for non-touch external drawing tablets & mouse.

**Reply by casept** [reply]
I appreciate your detailed reply, thank you.

I think I'll implement the node-based approach first, so I can concentrate on the actual ruler logic.

I don't like it from a UI perspective though, as I consider the gesture-based approach much more intuitive and less annoying than having to grab small handles.
